tiledb/sm/query/readers/attribute_order_validator.h

```diff
@@ -0,0 +1,608 @@
+        if (value_previous >= value) {
+          throw AttributeOrderValidatorStatusException(
+              "Attribute out of order");
+        }
+      } else {
+        if (value_previous <= value) {
+          throw AttributeOrderValidatorStatusException(
+              "Attribute out of order");
+        }
+      }
+    }
+
+    if (!val_data.max_validated_) {
+      // Get the min of the current fragment.
+      auto max_tile_idx = fragment_metadata[f]->tile_num() - 1;
+      auto value = fragment_metadata[f]->get_tile_max_as<AttributeType>(
+          attribute_name_, max_tile_idx);
+
+      // Get the previous value from the loaded tile.
+      auto rt = max_tile_to_compare_against(f);
+      const auto cell_idx =
+          non_empty_domain[1] -
+          index_dim->tile_coord_low(
+              rt->tile_idx() + frag_first_array_tile_idx[rt->frag_idx()],
+              dim_dom[0],
+              tile_extent) +
+          1;
+      AttributeType value_next =
+          rt->attribute_value<AttributeType>(attribute_name_, cell_idx);
+
+      // Validate the order.
+      if (increasing_data) {
+        if (value >= value_next) {
+          throw AttributeOrderValidatorStatusException(
+              "Attribute out of order");
+        }
+      } else {
+        if (value <= value_next) {
+          throw AttributeOrderValidatorStatusException(
+              "Attribute out of order");
+        }
+      }
+    }
+  }
+
+ private:
+  /* ********************************* */
+  /*       PRIVATE DECLARATIONS        */
+  /* ********************************* */
+
+  class AttributeOrderValidationData {
+   public:
+    AttributeOrderValidationData()
+        : min_validated_(false)
+        , max_validated_(false)
+        , min_frag_to_compare_to_(nullopt)
+        , max_frag_to_compare_to_(nullopt)
+        , min_tile_to_compare_to_(nullopt)
+        , max_tile_to_compare_to_(nullopt) {
+    }
+
+    /** Has the min has been validated already. */
+    bool min_validated_;
+
+    /** Has the max has been validated already. */
+    bool max_validated_;
+
+    /*
+     * Which fragment index to validate the min bound against.
+     *
+     * If it was possible to validate the value without looking at another
+     * fragment, this will be nullopt.
+     */
+    optional<uint64_t> min_frag_to_compare_to_;
+
+    /*
+     * Which fragment index to validate the max bound against.
+     *
+     * If it was possible to validate the value without looking at another
+     * fragment, this will be nullopt.
+     */
+    optional<uint64_t> max_frag_to_compare_to_;
+
+    /**
+     * Which tile we should compare data against for the min. The value is an
+     * index into `result_tiles_to_load_`.
+     */
+    optional<uint64_t> min_tile_to_compare_to_;
+
+    /**
+     * Which tile we should compare data against for the max. The value is an
+     * index into `result_tiles_to_load_`.
+     */
+    optional<uint64_t> max_tile_to_compare_to_;
```

Comment:
```suggestion
    /** Has the min has been validated already. */
    bool min_validated_{false};

    /** Has the max has been validated already. */
    bool max_validated_{false};

    /*
     * Which fragment index to validate the min bound against.
     *
     * If it was possible to validate the value without looking at another
     * fragment, this will be nullopt.
     */
    optional<uint64_t> min_frag_to_compare_to_{nullopt};

    /*
     * Which fragment index to validate the max bound against.
     *
     * If it was possible to validate the value without looking at another
     * fragment, this will be nullopt.
     */
    optional<uint64_t> max_frag_to_compare_to_{nullopt};

    /**
     * Which tile we should compare data against for the min. The value is an
     * index into `result_tiles_to_load_`.
     */
    optional<uint64_t> min_tile_to_compare_to_{nullopt};

    /**
     * Which tile we should compare data against for the max. The value is an
     * index into `result_tiles_to_load_`.
     */
    optional<uint64_t> max_tile_to_compare_to_ {nullopt};
```